File: packages/worker/src/journal.ts
import { appendFileSync, existsSync, mkdirSync, readFileSync } from 'node:fs';
import { dirname } from 'node:path';
import type { ParticleEvent } from '@particle/core';

export interface EventStore {
  load(): Promise<ParticleEvent[]>;
  append(events: ParticleEvent[]): Promise<void>;
}

/**
 * Phase-0 stand-in for the git ref store (P1.T3): an append-only NDJSON event
 * journal on disk. Same events, same fold — only the transport is temporary.
 */
export class Journal implements EventStore {
  constructor(private readonly path: string) {
    mkdirSync(dirname(path), { recursive: true });
  }

  async load(): Promise<ParticleEvent[]> {
    if (!existsSync(this.path)) return [];
    return readFileSync(this.path, 'utf8')
      .split('\n')
      .filter((line) => line.trim() !== '')
      .map((line) => JSON.parse(line) as ParticleEvent);
  }

  async append(events: ParticleEvent[]): Promise<void> {
    if (events.length === 0) return;
    appendFileSync(this.path, events.map((e) => JSON.stringify(e) + '\n').join(''));
  }
}

File: rust/particle-worker/src/journal.rs
use std::fs::{self, OpenOptions};
use std::io::Write;
use std::path::PathBuf;

use anyhow::{Context, Result};
use particle_core::ParticleEvent;

/// Phase-0 stand-in for the git ref store (P1.T3): an append-only NDJSON event
/// journal on disk, format-compatible with the TypeScript worker's journal.
pub struct Journal {
    path: PathBuf,
}

impl Journal {
    pub fn new(path: PathBuf) -> Result<Self> {
        if let Some(dir) = path.parent() {
            fs::create_dir_all(dir)?;
        }
        Ok(Journal { path })
    }

    pub fn load(&self) -> Result<Vec<ParticleEvent>> {
        if !self.path.exists() {
            return Ok(vec![]);
        }
        let raw = fs::read_to_string(&self.path)?;
        raw.lines()
            .filter(|line| !line.trim().is_empty())
            .map(|line| {
                serde_json::from_str(line).with_context(|| format!("bad journal line: {line}"))
            })
            .collect()
    }

    pub fn append(&self, events: &[ParticleEvent]) -> Result<()> {
        if events.is_empty() {
            return Ok(());
        }
        let mut file = OpenOptions::new()
            .create(true)
            .append(true)
            .open(&self.path)?;
        for event in events {
            serde_json::to_writer(&mut file, event)?;
            file.write_all(b"\n")?;
        }
        Ok(())
    }
}
